iroha/src/queue.rs

```diff
@@ -1,214 +1,161 @@
+                // MST case
+                old_tx
+                    .get_mut()
+                    .as_mut_inner_v1()
+                    .signatures
+                    .append(&mut tx.into_inner_v1().signatures);
+                return Ok(());
+            }
+            Entry::Vacant(entry) => entry,
+        };
+
+        if self.len.load() >= self.txs_in_queue {
+            return Err(tx);
         }
+
+        #[allow(clippy::expect_used)]
+        self.queue
+            .push(tx.hash())
+            .expect("Safe as we checked number of items just above");
+        let _ = self.len.fetch_add(1);
+        drop(entry.insert(tx));
+        Ok(())
     }
 
-    /// Gets at most `max_txs_in_block` number of transactions, but does not drop them out of the queue.
-    /// Drops only the transactions that have reached their TTL or are already in blockchain.
-    /// For MST transactions if on leader, waits for them to gather enough signatures before, showing them as output of this function.
+    /// Pops single transaction.
     ///
-    /// The reason for not dropping transaction when getting them, is that in the case of a view change this peer might become a leader,
-    /// or might need to froward transaction to the leader to check if the leader is not faulty.
-    /// If there is no view change and the block is commited then the transactions will simply drop because they are in a blockchain already.
-    #[allow(clippy::expect_used)]
-    pub fn get_pending_txs(&mut self, is_leader: bool) -> Vec<VersionedAcceptedTransaction> {
-        let mut output_txs = Vec::new();
-        let mut left_behind_txs = VecDeque::new();
-        let mut counter = self.max_txs_in_block;
-
-        while counter > 0 && !self.pending_tx_hash_queue.is_empty() {
-            let tx_hash = self
-                .pending_tx_hash_queue
-                .pop_front()
-                .expect("Unreachable, as queue not empty");
-            let tx = &self.pending_tx_by_hash[&tx_hash];
-
-            if tx.is_expired(self.tx_time_to_live) || tx.is_in_blockchain(&*self.wsv) {
-                drop(
-                    self.pending_tx_by_hash
-                        .remove(&tx_hash)
-                        .expect("Should always be present, as contained in queue"),
-                );
-                continue;
-            }
+    /// Records unsigned transaction in seen.
+    #[allow(clippy::unwrap_used, clippy::unwrap_in_result)]
+    fn pop<W: WorldTrait>(
+        &self,
+        is_leader: bool,
+        wsv: &WorldStateView<W>,
+        seen: &mut Vec<Hash>,
+    ) -> Option<VersionedAcceptedTransaction> {
+        let hash = self.queue.pop()?;
+        let tx = self.txs.get(&hash).unwrap();
+
+        if tx.is_expired(self.ttl) || tx.is_in_blockchain(wsv) {
+            drop(tx);
+            let _ = self.len.fetch_sub(1);
+            drop(self.txs.remove(&hash));
+            return self.pop(is_leader, wsv, seen);
```

Comment:
Recursion is very dangerous in the queue. As the queue can contain thousands of transactions.